etrago/tools/utilities.py

```diff
@@ -2973,3 +2981,260 @@ def manual_fixes_datamodel(etrago):
             ].index,
             inplace=True,
         )
+
+
+def select_elec_network(etrago, apply_on="grid_model"):
+    """
+    Creates networks to be used on the clustering based on settings specified
+    in the args.
+
+    Parameters
+    ----------
+    etrago : Etrago
+        An instance of the Etrago class
+    apply_on: str
+        gives information about the objective of the output network. If
+        "grid_model" is provided, the value assigned in the args for
+        ["network_clustering"]["cluster_foreign_AC""] will define if the
+        foreign buses will be included in the network. if "market_model" is
+        provided, foreign buses will be always included.
+
+    Returns
+    -------
+    Tuple containing:
+        elec_network : pypsa.Network
+            Contains the electric network
+        n_clusters : int
+            number of clusters used in the clustering process.
+        area_network : pypsa.Network
+            Contains the electric network in the area of interest defined in
+            network_clustering - interest_area.
+    """
+    settings = etrago.args["network_clustering"]
+
+    if apply_on in ["grid_model", "grid_model-ehv"]:
+        # Find buses in the area that should not be clustered
+        buses_area = find_buses_area(etrago, "AC")
+
+        elec_network_buses = etrago.network.buses[
+            (~etrago.network.buses.index.isin(buses_area))
+            & (etrago.network.buses.carrier == "AC")
+        ].index
+        if apply_on == "grid_model-ehv":
+            n_clusters = pd.NA
+        # Exclude foreign buses when set to don't include them in clustering
+        elif settings["cluster_foreign_AC"]:
+            n_clusters = settings["n_clusters_AC"]
+        else:
+            foreign_buses = etrago.network.buses[
+                (etrago.network.buses.country != "DE")
+                & (etrago.network.buses.carrier == "AC")
+            ]
+
+            num_neighboring_country = len(
+                foreign_buses[
+                    foreign_buses.index.isin(etrago.network.loads.bus)
+                ]
+            )
+
+            elec_network_buses = elec_network_buses[
+                ~elec_network_buses.isin(foreign_buses.index)
+            ]
+            if settings["n_clusters_AC"] is False:
+                n_clusters = False
+            else:
+                n_clusters = (
+                    settings["n_clusters_AC"] - num_neighboring_country
+                )
+                assert (
+                    n_clusters > 1
+                ), f"""'n_clusters_AC' must be greater than the number of
+                foreign countries({num_neighboring_country})"""
+
+        elec_network = network_based_on_buses(
+            etrago.network, elec_network_buses
+        )
+        area_network = network_based_on_buses(etrago.network, buses_area)
+
+    elif apply_on == "market_model":
+        elec_network_buses = etrago.network_tsa.buses[
```

Comment:
@CarlosEpia or @ClaraBuettner, I discovered that n_clusters is not assigned in this elif. Not sure if this was done in this branch or not, so maybe it's not an issue that should be fixed here.

Reply:
This function is defined in cluster.electrical.py on `dev`. In that function, n_clusters is defined for all settings and the clustering works fine. So I guess it is related to this branch. Could you have a look at it @CarlosEpia?

Reply:
@birgits good catch! Thanks for the warning. Problem solved in https://github.com/openego/eTraGo/commit/cec6836f640fda2a50adb113f550452122673af5.